File: moling-server/app/dao/dynamic_layer_dao.py
"""墨灵 (Moling) — DynamicLayer DAO."""

from __future__ import annotations

from typing import Optional

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.dao.base_dao import BaseDAO
from app.models.dynamic_layer import DynamicLayer


class DynamicLayerDAO(BaseDAO[DynamicLayer]):
    """Data access for DynamicLayer (dynamic context per chapter)."""

    def __init__(self) -> None:
        super().__init__(DynamicLayer)

    async def get_by_chapter(
        self,
        db: AsyncSession,
        chapter_id: str,
    ) -> Optional[DynamicLayer]:
        """Get the dynamic layer for a specific chapter."""
        stmt = (
            select(DynamicLayer)
            .where(DynamicLayer.chapter_id == chapter_id)
            .limit(1)
        )
        result = await db.execute(stmt)
        return result.scalar_one_or_none()

    async def get_recent_by_project(
        self,
        db: AsyncSession,
        project_id: str,
        limit: int = 5,
    ) -> list[DynamicLayer]:
        """Get the most recent dynamic layers for a project."""
        stmt = (
            select(DynamicLayer)
            .where(DynamicLayer.project_id == project_id)
            .order_by(DynamicLayer.id.desc())
            .limit(limit)
        )
        result = await db.execute(stmt)
        return list(result.scalars().all())

    async def get_latest_by_project(
        self,
        db: AsyncSession,
        project_id: str,
    ) -> DynamicLayer | None:
        """Get the single most recent dynamic layer for a project."""
        layers = await self.get_recent_by_project(db, project_id, limit=1)
        return layers[0] if layers else None

    async def get_health_check_history(
        self,
        db: AsyncSession,
        project_id: int,
        limit: int = 20,
        *,
        start_chapter: int | None = None,
        end_chapter: int | None = None,
    ) -> list[dict]:
        """Get health check history with chapter numbers for a project.

        Returns list of dicts with keys: health_check, chapter_number, checked_at

        Optionally filter by chapter_number range: start_chapter <= n < end_chapter.
        """
        from app.models.chapter import Chapter

        stmt = (
            select(
                DynamicLayer.health_check,
                Chapter.chapter_number,
                DynamicLayer.created_at.label("checked_at"),
            )
            .join(Chapter, DynamicLayer.chapter_id == Chapter.id)
            .where(DynamicLayer.project_id == project_id)
            .where(DynamicLayer.health_check.isnot(None))
            .order_by(DynamicLayer.created_at.desc())
            .limit(limit)
        )
        if start_chapter is not None:
            stmt = stmt.where(Chapter.chapter_number >= start_chapter)
        if end_chapter is not None:
            stmt = stmt.where(Chapter.chapter_number < end_chapter)

        result = await db.execute(stmt)
        rows = result.all()
        return [
            {
                "health_check": row.health_check,
                "chapter_number": row.chapter_number,
                "checked_at": row.checked_at,
            }
            for row in rows
        ]


# Singleton
dynamic_layer_dao = DynamicLayerDAO()

File: moling-server-rs/crates/moling-db/src/dao/dynamic_layer_dao.rs
//! Dynamic Layer DAO — dynamic context layer CRUD and health-check history.

use moling_core::error::{AppError, AppResult};
use sea_orm::{ActiveModelTrait, ColumnTrait, DatabaseConnection, EntityTrait, QueryFilter, QueryOrder, QuerySelect};
use sea_orm::prelude::DateTimeUtc;

use crate::entities::dynamic_layer::{self, ActiveModel, Entity as DynamicLayer, Model as DynamicLayerModel};

/// Dynamic layer data access object.
#[derive(Clone, Default)]
pub struct DynamicLayerDao;

impl DynamicLayerDao {
    /// Find a dynamic layer by ID.
    pub async fn find_by_id(
        &self,
        db: &DatabaseConnection,
        id: &str,
    ) -> AppResult<Option<DynamicLayerModel>> {
        DynamicLayer::find_by_id(id.to_owned())
            .one(db)
            .await
            .map_err(|e| {
                tracing::error!(%id, "DynamicLayer: database error finding by id: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// Get the dynamic layer for a specific chapter.
    pub async fn find_by_chapter(
        &self,
        db: &DatabaseConnection,
        chapter_id: &str,
    ) -> AppResult<Option<DynamicLayerModel>> {
        DynamicLayer::find()
            .filter(dynamic_layer::Column::ChapterId.eq(chapter_id))
            .one(db)
            .await
            .map_err(|e| {
                tracing::error!(%chapter_id, "DynamicLayer: database error finding by chapter: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// Get the most recent dynamic layers for a project.
    pub async fn list_recent_by_project(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
        limit: u64,
    ) -> AppResult<Vec<DynamicLayerModel>> {
        DynamicLayer::find()
            .filter(dynamic_layer::Column::ProjectId.eq(project_id))
            .order_by_desc(dynamic_layer::Column::Id)
            .limit(limit)
            .all(db)
            .await
            .map_err(|e| {
                tracing::error!(project_id, "DynamicLayer: database error listing recent: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// Get the single most recent dynamic layer for a project.
    pub async fn find_latest_by_project(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
    ) -> AppResult<Option<DynamicLayerModel>> {
        let layers = self.list_recent_by_project(db, project_id, 1).await?;
        Ok(layers.into_iter().next())
    }

    /// Get health check history joined with chapter numbers for a project.
    /// Returns tuples of (health_check_json, chapter_number, checked_at).
    pub async fn get_health_check_history(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
        limit: u64,
        start_chapter: Option<i32>,
        end_chapter: Option<i32>,
    ) -> AppResult<Vec<(serde_json::Value, i32, DateTimeUtc)>> {
        use crate::entities::chapter;
        use sea_orm::{JoinType, QuerySelect, RelationTrait};

        let mut query = DynamicLayer::find()
            .select_only()
            .column(dynamic_layer::Column::HealthCheck)
            .column_as(chapter::Column::ChapterNumber, "chapter_number")
            .column_as(dynamic_layer::Column::CreatedAt, "checked_at")
            .join_rev(
                JoinType::InnerJoin,
                chapter::Relation::DynamicLayer.def().rev(),
            )
            .filter(dynamic_layer::Column::ProjectId.eq(project_id))
            .filter(dynamic_layer::Column::HealthCheck.is_not_null())
            .order_by_desc(dynamic_layer::Column::CreatedAt)
            .limit(limit);

        if let Some(start) = start_chapter {
            query = query.filter(chapter::Column::ChapterNumber.gte(start));
        }
        if let Some(end) = end_chapter {
            query = query.filter(chapter::Column::ChapterNumber.lt(end));
        }

        query
            .into_tuple::<(serde_json::Value, i32, DateTimeUtc)>()
            .all(db)
            .await
            .map_err(|e| {
                tracing::error!(project_id, "DynamicLayer: database error fetching health check history: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// Create a new dynamic layer.
    pub async fn create(
        &self,
        db: &DatabaseConnection,
        model: ActiveModel,
    ) -> AppResult<DynamicLayerModel> {
        model.insert(db).await.map_err(|e| {
            tracing::error!("DynamicLayer: database error creating: {e}");
            AppError::internal("Database insert failed")
        })
    }

    /// Update an existing dynamic layer.
    pub async fn update(
        &self,
        db: &DatabaseConnection,
        model: ActiveModel,
    ) -> AppResult<DynamicLayerModel> {
        model.update(db).await.map_err(|e| {
            tracing::error!("DynamicLayer: database error updating: {e}");
            AppError::internal("Database update failed")
        })
    }
}
